Collective lift and shift updates

jonrandahl opened the pull request:
This PR includes the following updates and resolves the relevent linked issues:

- Added a cli recommendation for running the `sr-manager` locally;
  resolved a markdown linting issue with using HTML in markdown; as well as
  fixed a simple typo in the `localhost:3001` port example
- Refactors the elapsed time calculated for API requests to be resolved as
  microseconds rather than milliseconds. This is to improve the reporting of the
  elapsed time in the system tooling logs. [#104](https://github.com/epimorphics/hmlr-linked-data/issues/104)/[#111](https://github.com/epimorphics/hmlr-linked-data/issues/111)
- Minor text changes to the `Gemfile` to include instructions for running
  Epimorphics specific gems locally during the development of those gems.
- Updated the production `data_services_api` gem version to be at least
  the current version`~>1.3.3` (this is to cover out of sync release versions)
- Updated the production `lr_common_styles` gem version to be at least the
  current version `~>1.9.1` (this is to cover out of sync release versions)
- Refactored better guards in `entrypoint.sh` to ensure the required env
  vars are set accordingly or deployment will fail noisily.
- Refactored the error messages displayed to be semantically formatted as
  well as incorporated ALL lines for the returned message to be held in the
  variable instead of being possibly displayed outside of the intended context.

jonrandahl committed:
build: updated version cadence 

Refactored the version cadence creation to include a SUFFIX value if provided; otherwise no SUFFIX is included in the version number.

jonrandahl committed:
style: updated Rubocop rule

Allows slightly longer than default length of method names which removes the need to override as comments

jonrandahl committed:
fix: Swap out Time.zone.now for Process.clock_gettime

Following community guidance for measuring elapsed time in rails the logging mechanism has been changed to use the CPU time instead of RealTime. This also has the added benefit to return the results in the requested `microseconds` unit without further mathematical manipulation.

jonrandahl committed:
docs: Updated docker identity

Updated as per the instructions

jonrandahl committed:
build: Setting Epimorphics gems based on environment 

Refined text changes to the `Gemfile` to include instructions for running Epimorphics specific gems locally during the development of those gems. Updated the production environment gem versions to be at least the current version to cover out of sync release versions.

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
build: No default values in production

In `Production` environment no default values are passed on the basis that missing configuration options represent a category of bug, and in that case the deployment should fail fast and noisily.

https://github.com/epimorphics/ijd-transfer-notes/blob/main/rails-app-pattern.md#environment-variables-and-run-time-configuration

jonrandahl committed:
feat: static asset cache headers

This sets the cache control headers for HMLR apps to be public and cacheable in the production environment. 
Standard Reports uses a time limit of 5 minutes (300 seconds)

https://github.com/epimorphics/hmlr-linked-data/issues/114

jonrandahl committed:
refactor: `RAILS_RELATIVE_URL_ROOT` renamed to `APPLICATION_PATH`

Following lengthy internal discussions the global env variable `RAILS_RELATIVE_URL_ROOT` has been renamed to `APPLICATION_PATH` to be more clear on it's use; needs to be monitored once deployed. Also includes better guards in `entrypoint.sh` to ensure the required env vars are set accordingly or deployment will fail noisily.

jonrandahl committed:
docs: Updated CHANGELOG

Also includes minor update to layout format for older entries

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
ci: resolving gem locations

refactored the gem paths to use the remote Epimorphics repository locations, also includes the regenerated gemfile.lock reflecting as such.

jonrandahl committed:
docs: Updated README

Added a cli recommendation for running the `sr-manager` locally; resolved a markdown linting issue with using HTML in markdown; as well as fixed a simple typo

jonrandahl committed:
docs: Updated CHANGELOG

der commented on CHANGELOG.md:
```diff
@@ -1,10 +1,31 @@
+- (Jon) Updated the production `json_rails_logger` gem version to be at least the
+  current version `~>1.3.5` (this is to cover out of sync release versions)
+- (Jon) Updated the production `lr_common_styles` gem version to be at least the
+  current version `~>1.9.1` (this is to cover out of sync release versions)
+- (Jon) Renamed the global env variable `RAILS_RELATIVE_URL_ROOT` to
```
If that definitely works and you find it more intuitive then that's fine.

Just wanted to note that RAILS_RELATIVE_URL_ROOT is a conventional name that you see in e.g. github issues and some of the tool chain, notably webpacker is aware if it. Given rails "convention over configuration" mantra this leaves me a little nervous :)

jonrandahl replied:
@der I've resolved the codebase to use the requested `APPLICATION_ROOT` variable name for the build process and the `RAILS_RELATIVE_URL_ROOT` for the app via the `entrypoint.sh` code. 🙏

der reviewed (approved):
Looks fine to me other than unfounded fear, uncertainty and doubt over the renaming of RAILS_RELATIVE_URL_ROOT.

jonrandahl committed:
fix: resolve `RAILS_RELATIVE_URL_ROOT` env var

This resolves that Rails would want the `RAILS_RELATIVE_URL_ROOT` env var set, while using `APPLICATION_ROOT` is clearer in the `entrypoint.sh` code - this also updates the readme’s to reflect this change

jonrandahl committed:
docs: Updated CHANGELOG

Fixed typo for `APPLICATION_ROOT` env var

jonrandahl committed:
Merge branch 'dev-infrastructure' into spike/collective-lift-and-shift-updates

jonrandahl committed:
fix: resolve `RAILS_RELATIVE_URL_ROOT` env var

sets environment variables for development environment

jonrandahl committed:
docs: Updated README

bogdanadrianmarc reviewed (approved):
Changes look good from a frontend perspective 👍

jonrandahl committed:
Merge remote-tracking branch 'origin/spike/collective-lift-and-shift-updates' into spike/collective-lift-and-shift-updates

der reviewed (approved):
:+1:

jonrandahl committed:
Merge branch 'dev-infrastructure' into spike/collective-lift-and-shift-updates

jonrandahl committed:
build: cleanup of build files for LR apps

Implements unification approach to Make, Docker and other associated build files to ensure the image process completes as expected

jonrandahl committed:
refactor: mirrored config setting across all apps

`config.assets.precompile` and `config.assets.version` have moved to `config/initializers/assets.rb`

jonrandahl committed:
build: unified environment configurations across all apps

Updated the matching configuration settings on each environment configuration file to uniify the settings for all apps

jonrandahl committed:
Merge remote-tracking branch 'origin/spike/collective-lift-and-shift-updates' into spike/collective-lift-and-shift-updates

andrew-pickin-epi reviewed (changes requested):
Please fix as pre LR Landing then I will review

der reviewed (approved):
As per others. I don't understand using `SCRIPT_NAME` and `APPLICATION_ROOT` instead of just `RAILS_RELATIVE_URL_ROOT` but once the pattern is resolved on the others assume it'll be applied here so approving.

jonrandahl committed:
build: latest updates for env specific variables and fallbacks

jonrandahl committed:
Merge remote-tracking branch 'origin/dev-infrastructure' into spike/collective-lift-and-shift-updates

jonrandahl committed:
build: entrypoint.sh cleanup

jonrandahl committed:
build: gemfile lock updates

includes latest epimorphics gems and other dependency updates

jonrandahl committed:
build: corresponding gem set

Updates to comments as per the other HMLR apps

jonrandahl committed:
fix: typo on API_SERVICE_URL default port

Updated to use `:8888` as per recommended

andrew-pickin-epi commented on CHANGELOG.md:
```diff
@@ -1,10 +1,31 @@
+- (Jon) Updated the production `lr_common_styles` gem version to be at least the
+  current version `~>1.9.1` (this is to cover out of sync release versions)
+- (Jon) Renamed the global env variable `RAILS_RELATIVE_URL_ROOT` to
+  `APPLICATION_ROOT` to be more clear on it's use in the `entrypoint.sh` code.
```
This change log would have us still using APPLICATION_ROOT, which I don't think is correct.

andrew-pickin-epi replied:
Summary of this PR still refers to APPLICATION_ROOT also.

jonrandahl replied:
Removed that line from changelog, thank you!

andrew-pickin-epi commented on Makefile:
```diff
@@ -26,7 +26,7 @@ IMAGE?=${NAME}/${STAGE}
 
 GITHUB_TOKEN=.github-token
-BUNDLE_CFG=.bundle/config
+BUNDLE_CFG=${HOME}/.bundle/config
```
This shouldn't be ${HOME}.

jonrandahl replied:
Removed `${HOME}/` from path!

jonrandahl committed:
build: cleaned BUNDLE_CONFIG

Removed `${HOME}/` from path

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
docs: Rewrote README

Includes further clarification on running the app individually, or as part of the HMLR suite of apps; alongside using Docker containers and accessing the Standard Reports Manager locally.
